ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java

```diff
@@ -311,13 +336,22 @@ protected void validateLogEntry(LogEntryProto entry) {
     if (entry.hasMetadataEntry()) {
       return;
     }
+    long latestSnapshotIndex = getSnapshotIndex();
     TermIndex lastTermIndex = getLastEntryTermIndex();
     if (lastTermIndex != null) {
+      long lastIndex = lastTermIndex.getIndex() > latestSnapshotIndex ?
+          lastTermIndex.getIndex() : latestSnapshotIndex;
       Preconditions.assertTrue(entry.getTerm() >= lastTermIndex.getTerm(),
           "Entry term less than RaftLog's last term: %d, entry: %s", lastTermIndex.getTerm(), entry);
-      Preconditions.assertTrue(entry.getIndex() == lastTermIndex.getIndex() + 1,
-          "Difference between entry index and RaftLog's last index %d greater than 1, entry: %s",
-          lastTermIndex.getIndex(), entry);
+      Preconditions.assertTrue(entry.getIndex() == lastIndex + 1,
+          "Difference between entry index and RaftLog's last index %d (or snapshot index %d) " +
+              "is greater than 1, entry: %s",
+          lastTermIndex.getIndex(), latestSnapshotIndex, entry);
+    } else {
+      Preconditions.assertTrue(entry.getIndex() == latestSnapshotIndex + 1,
+          "Difference between entry index and RaftLog's latest snapshot index is greater than 1 " +
```

Comment:
We are missing a %d for latestSnapshotIndex.

Reply:
Thanks Lokesh. Updated.